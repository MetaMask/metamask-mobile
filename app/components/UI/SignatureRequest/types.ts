export interface MessageInfo {
  origin: string;
  type: string;
}

export interface PageMeta {
  analytics?: {
    request_platform: string;
    request_source: string;
  };
  icon?: string;
  title: string;
  url: string;
}

export interface MessageParams {
  data: string;
  from: string;
  metamaskId: string;
  meta?: PageMeta;
  origin: string;
}
