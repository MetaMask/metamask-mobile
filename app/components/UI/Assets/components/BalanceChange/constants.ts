export const FORMATTED_VALUE_PRICE_TEST_ID = 'formatted-value-price-test-id';
export const FORMATTED_PERCENTAGE_TEST_ID = 'formatted-percentage-test-id';
