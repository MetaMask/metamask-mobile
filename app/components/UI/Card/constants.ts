import { ethers } from 'ethers';
import balanceScannerAbi from './sdk/balanceScannerAbi.json';

export const BALANCE_SCANNER_ABI =
  balanceScannerAbi as ethers.ContractInterface;
export const ARBITRARY_ALLOWANCE = 100000000000;
