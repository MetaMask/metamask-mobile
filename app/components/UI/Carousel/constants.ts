import { ImageSourcePropType } from 'react-native';
import { strings } from '../../../../locales/i18n';
import { CarouselSlide, SlideId } from './types';
import { WalletViewSelectorsIDs } from '../../../../e2e/selectors/wallet/WalletView.selectors';
import {
  createBuyNavigationDetails,
  createSellNavigationDetails,
} from '../../UI/Ramp/routes/utils';
import Routes from '../../../constants/navigation/Routes';
import cardImage from '../../../images/banners/banner_image_card.png';
import fundImage from '../../../images/banners/banner_image_fund.png';
import cashoutImage from '../../../images/banners/banner_image_cashout.png';
import aggregatedImage from '../../../images/banners/banner_image_aggregated.png';

export const PREDEFINED_SLIDES: CarouselSlide[] = [
  {
    id: 'card',
    title: strings('banner.card.title'),
    description: strings('banner.card.subtitle'),
    undismissable: false,
    navigation: { type: 'url', href: 'https://portfolio.metamask.io/card' },
    testID: WalletViewSelectorsIDs.CAROUSEL_FIRST_SLIDE,
    testIDTitle: WalletViewSelectorsIDs.CAROUSEL_FIRST_SLIDE_TITLE,
    testIDCloseButton: WalletViewSelectorsIDs.CAROUSEL_FIRST_SLIDE_CLOSE_BUTTON,
  },
  {
    id: 'fund',
    title: strings('banner.fund.title'),
    description: strings('banner.fund.subtitle'),
    undismissable: false,
    navigation: {
      type: 'function',
      navigate: () => createBuyNavigationDetails(),
    },
    testID: WalletViewSelectorsIDs.CAROUSEL_SECOND_SLIDE,
    testIDTitle: WalletViewSelectorsIDs.CAROUSEL_SECOND_SLIDE_TITLE,
    testIDCloseButton:
      WalletViewSelectorsIDs.CAROUSEL_SECOND_SLIDE_CLOSE_BUTTON,
  },
  {
    id: 'cashout',
    title: strings('banner.cashout.title'),
    description: strings('banner.cashout.subtitle'),
    undismissable: false,
    navigation: {
      type: 'function',
      navigate: () => createSellNavigationDetails(),
    },
    testID: WalletViewSelectorsIDs.CAROUSEL_THIRD_SLIDE,
    testIDTitle: WalletViewSelectorsIDs.CAROUSEL_THIRD_SLIDE_TITLE,
    testIDCloseButton: WalletViewSelectorsIDs.CAROUSEL_THIRD_SLIDE_CLOSE_BUTTON,
  },
  {
    id: 'aggregated',
    title: strings('banner.aggregated.title'),
    description: strings('banner.aggregated.subtitle'),
    undismissable: false,
    navigation: {
      type: 'route',
      route: Routes.ONBOARDING.GENERAL_SETTINGS,
      navigationStack: Routes.SETTINGS_VIEW,
    },
    testID: WalletViewSelectorsIDs.CAROUSEL_FOURTH_SLIDE,
    testIDTitle: WalletViewSelectorsIDs.CAROUSEL_FOURTH_SLIDE_TITLE,
    testIDCloseButton:
      WalletViewSelectorsIDs.CAROUSEL_FOURTH_SLIDE_CLOSE_BUTTON,
  },
];

export const BANNER_IMAGES: Record<SlideId, ImageSourcePropType> = {
  card: cardImage,
  fund: fundImage,
  cashout: cashoutImage,
  aggregated: aggregatedImage,
};
