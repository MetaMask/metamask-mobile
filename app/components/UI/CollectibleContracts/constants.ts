export const RefreshTestId = 'refreshControl';
export const SpinnerTestId = 'spinner';
