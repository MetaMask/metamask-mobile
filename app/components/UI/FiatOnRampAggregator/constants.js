import { Icon } from './components/PaymentIcon';

export const CHAIN_ID_NETWORKS = {
	1: 'ethereum',
	10: 'optimism',
	56: 'bsc',
	137: 'polygon',
	200: 'arbitrum',
	43114: 'avalanche',
};

export const PAYMENT_METHOD_ICON = {
	'/payments/debit-credit-card': Icon.Card,
	'/payments/bank-account': Icon.Bank,
	'/payments/apple-pay': Icon.Apple,
};
