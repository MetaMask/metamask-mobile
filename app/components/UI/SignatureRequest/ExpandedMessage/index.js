import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
  StyleSheet,
  View,
  Text,
  TouchableOpacity,
  ScrollView,
  TouchableWithoutFeedback,
} from 'react-native';
import { fontStyles, baseStyles } from '../../../../styles/common';
import WebsiteIcon from '../../WebsiteIcon';
import Ionicons from 'react-native-vector-icons/Ionicons';
import { strings } from '../../../../../locales/i18n';
import Device from '../../../../util/device';
import { getHost } from '../../../../util/browser';
import { ThemeContext, mockTheme } from '../../../../util/theme';

const createStyles = (colors) =>
  StyleSheet.create({
    expandedRoot: {
      backgroundColor: colors.background.default,
      minHeight: Device.isIos() ? '70%' : '80%',
      borderTopLeftRadius: 20,
      borderTopRightRadius: 20,
      padding: 24,
      paddingBottom: Device.isIphoneX() ? 44 : 24,
    },

    expandedMessageHeader: {
      width: '100%',
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      marginBottom: 20,
    },
    arrowIcon: {
      ...baseStyles.flexGrow,
      color: colors.icon.muted,
    },
    iconHidden: {
      ...baseStyles.flexGrow,
    },
    messageLabelTextExpanded: {
      ...baseStyles.flexGrow,
      textAlign: 'center',
      ...fontStyles.bold,
      fontSize: 16,
      color: colors.text.default,
    },
    messageIntroWrapper: {
      alignItems: 'center',
      marginBottom: 20,
    },
    domainLogo: {
      width: 40,
      height: 40,
      borderRadius: 20,
      marginBottom: 20,
    },
    messageFromLabel: {
      textAlign: 'center',
      ...fontStyles.bold,
      fontSize: 16,
      color: colors.text.default,
    },
    scrollView: {
      ...baseStyles.flexGrow,
    },
  });

/**
 * Component that supports eth_signTypedData and eth_signTypedData_v3
 */
export default class ExpandedMessage extends PureComponent {
  static propTypes = {
    /**
     * Object containing current page title and url
     */
    currentPageInformation: PropTypes.object,
    /**
     * Renders the message based on its type (parent)
     */
    renderMessage: PropTypes.func,
    /**
     * Expands the message box on press.
     */
    toggleExpandedMessage: PropTypes.func,
  };

  render() {
    const { currentPageInformation, renderMessage, toggleExpandedMessage } =
      this.props;
    const url = currentPageInformation.url;
    const title = getHost(url);
    const colors = this.context.colors || mockTheme.colors;
    const styles = createStyles(colors);

    return (
      <View style={styles.expandedRoot}>
        <TouchableOpacity
          style={styles.expandedMessageHeader}
          onPress={toggleExpandedMessage}
        >
          <Ionicons
            name={'ios-arrow-back'}
            size={30}
            style={styles.arrowIcon}
          />
          <Text style={styles.messageLabelTextExpanded}>
            {strings('signature_request.message')}
          </Text>
          <View style={styles.iconHidden} />
        </TouchableOpacity>
        <View style={styles.messageIntroWrapper}>
          <WebsiteIcon style={styles.domainLogo} title={title} url={url} />
          <Text style={styles.messageFromLabel}>
            {strings('signature_request.message_from')} {title}
          </Text>
        </View>
        <ScrollView style={styles.scrollView}>
          <TouchableWithoutFeedback>
            <View>{renderMessage()}</View>
          </TouchableWithoutFeedback>
        </ScrollView>
      </View>
    );
  }
}

ExpandedMessage.contextType = ThemeContext;
