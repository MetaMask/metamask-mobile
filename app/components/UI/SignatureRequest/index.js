import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, View, Text, TouchableOpacity } from 'react-native';
import { fontStyles } from '../../../styles/common';
import { getHost } from '../../../util/browser';
import { strings } from '../../../../locales/i18n';
import { connect } from 'react-redux';
import Ionicons from 'react-native-vector-icons/Ionicons';
import WebsiteIcon from '../WebsiteIcon';
import ActionView from '../ActionView';
import AccountInfoCard from '../AccountInfoCard';
import TransactionHeader from '../TransactionHeader';
import WarningMessage from '../../Views/SendFlow/WarningMessage';
import Device from '../../../util/device';
import Analytics from '../../../core/Analytics/Analytics';
import { MetaMetricsEvents } from '../../../core/Analytics';
import { ThemeContext, mockTheme } from '../../../util/theme';
import withQRHardwareAwareness from '../QRHardware/withQRHardwareAwareness';
import QRSigningDetails from '../QRHardware/QRSigningDetails';

const createStyles = (colors) =>
  StyleSheet.create({
    root: {
      backgroundColor: colors.background.default,
      paddingTop: 24,
      minHeight: '90%',
      borderTopLeftRadius: 20,
      borderTopRightRadius: 20,
      paddingBottom: Device.isIphoneX() ? 20 : 0,
    },
    expandedHeight2: {
      minHeight: '90%',
    },
    expandedHeight1: {
      minHeight: '90%',
    },
    signingInformation: {
      alignItems: 'center',
      marginVertical: 24,
    },
    domainLogo: {
      width: 40,
      height: 40,
      marginRight: 8,
      borderRadius: 20,
    },
    messageColumn: {
      width: '75%',
      justifyContent: 'space-between',
    },
    warningLink: {
      ...fontStyles.normal,
      color: colors.primary.default,
      textAlign: 'center',
      paddingHorizontal: 10,
      paddingBottom: 10,
      textDecorationLine: 'underline',
    },
    signText: {
      ...fontStyles.bold,
      fontSize: 20,
      textAlign: 'center',
      color: colors.text.default,
    },
    messageLabelText: {
      ...fontStyles.bold,
      fontSize: 16,
      marginBottom: 4,
      color: colors.text.default,
    },
    readMore: {
      color: colors.primary.default,
      fontSize: 14,
      ...fontStyles.bold,
    },
    warningWrapper: {
      width: '100%',
      paddingHorizontal: 24,
      paddingTop: 24,
    },
    actionViewChild: {
      paddingHorizontal: 24,
    },
    accountInfoCardWrapper: {
      marginBottom: 20,
      width: '100%',
    },
    children: {
      alignSelf: 'center',
      flexDirection: 'row',
      justifyContent: 'flex-start',
      width: '100%',
      borderWidth: 1,
      borderColor: colors.border.default,
      borderRadius: 10,
      padding: 16,
    },
    arrowIconWrapper: {
      flexGrow: 1,
      alignItems: 'flex-end',
    },
    arrowIcon: {
      color: colors.icon.muted,
    },
  });

/**
 * PureComponent that renders scrollable content inside signature request user interface
 */
class SignatureRequest extends PureComponent {
  static propTypes = {
    /**
     * Object representing the navigator
     */
    navigation: PropTypes.object,
    /**
     * Callback triggered when this message signature is rejected
     */
    onCancel: PropTypes.func,
    /**
     * Callback triggered when this message signature is approved
     */
    onConfirm: PropTypes.func,
    /**
     * Content to display above the action buttons
     */
    children: PropTypes.node,
    /**
     * Object containing current page title and url
     */
    currentPageInformation: PropTypes.object,
    /**
     * String representing signature type
     */
    type: PropTypes.string,
    /**
     * String representing the selected network
     */
    networkType: PropTypes.string,
    /**
     * Whether it should display the warning message
     */
    showWarning: PropTypes.bool,
    /**
     * Whether it should render the expand arrow icon
     */
    truncateMessage: PropTypes.bool,
    /**
     * Expands the message box on press.
     */
    toggleExpandedMessage: PropTypes.func,
    /**
     * Active address of account that triggered signing.
     */
    fromAddress: PropTypes.string,
    isSigningQRObject: PropTypes.bool,
    QRState: PropTypes.object,
  };

  /**
   * Calls trackCancelSignature and onCancel callback
   */
  onCancel = () => {
    this.props.onCancel();
    Analytics.trackEventWithParameters(
      MetaMetricsEvents.TRANSACTIONS_CANCEL_SIGNATURE,
      this.getTrackingParams(),
    );
  };

  /**
   * Calls trackConfirmSignature and onConfirm callback
   */
  onConfirm = () => {
    this.props.onConfirm();
    Analytics.trackEventWithParameters(
      MetaMetricsEvents.TRANSACTIONS_CONFIRM_SIGNATURE,
      this.getTrackingParams(),
    );
  };

  /**
   * Returns corresponding tracking params to send
   *
   * @return {object} - Object containing network and functionType
   */
  getTrackingParams = () => {
    const { type, networkType } = this.props;
    return {
      network: networkType,
      functionType: type,
    };
  };

  goToWarning = () => {
    this.props.onCancel();
    this.props.navigation.navigate('Webview', {
      screen: 'SimpleWebview',
      params: {
        url: 'https://metamask.zendesk.com/hc/en-us/articles/360015488751',
        title: 'metamask.zendesk.com',
      },
    });
  };

  getStyles = () => {
    const colors = this.context.colors || mockTheme.colors;
    return createStyles(colors);
  };

  renderWarning = () => {
    const styles = this.getStyles();

    return (
      <Text>
        {strings('signature_request.eth_sign_warning')}
        {` `}
        <Text style={styles.warningLink}>
          {strings('signature_request.learn_more')}
        </Text>
      </Text>
    );
  };

  renderActionViewChildren = () => {
    const {
      children,
      currentPageInformation,
      truncateMessage,
      toggleExpandedMessage,
      fromAddress,
    } = this.props;
    const styles = this.getStyles();
    const url = currentPageInformation.url;
    const title = getHost(url);
    const arrowIcon = truncateMessage ? this.renderArrowIcon() : null;
    return (
      <View style={styles.actionViewChild}>
        <View style={styles.accountInfoCardWrapper}>
          <AccountInfoCard operation="signing" fromAddress={fromAddress} />
        </View>
        <TouchableOpacity
          style={styles.children}
          onPress={truncateMessage ? toggleExpandedMessage : null}
        >
          <WebsiteIcon style={styles.domainLogo} title={title} url={url} />
          <View style={styles.messageColumn}>
            <Text style={styles.messageLabelText}>
              {strings('signature_request.message')}:
            </Text>
            {children}
            {truncateMessage ? (
              <Text style={styles.readMore}>
                {strings('signature_request.read_more')}
              </Text>
            ) : null}
          </View>
          <View style={styles.arrowIconWrapper}>{arrowIcon}</View>
        </TouchableOpacity>
      </View>
    );
  };

  renderArrowIcon = () => {
    const styles = this.getStyles();

    return (
      <View style={styles.arrowIconWrapper}>
        <Ionicons
          name={'ios-arrow-forward'}
          size={20}
          style={styles.arrowIcon}
        />
      </View>
    );
  };

  renderSignatureRequest() {
    const { showWarning, currentPageInformation, type } = this.props;
    let expandedHeight;
    const styles = this.getStyles();

    if (Device.isMediumDevice()) {
      expandedHeight = styles.expandedHeight2;
      if (type === 'ethSign') {
        expandedHeight = styles.expandedHeight1;
      }
    }
    return (
      <View style={[styles.root, expandedHeight]}>
        <ActionView
          cancelTestID={'request-signature-cancel-button'}
          confirmTestID={'request-signature-confirm-button'}
          cancelText={strings('signature_request.cancel')}
          confirmText={strings('signature_request.sign')}
          onCancelPress={this.onCancel}
          onConfirmPress={this.onConfirm}
          confirmButtonMode="sign"
        >
          <View>
            <TransactionHeader
              currentPageInformation={currentPageInformation}
              type={type}
            />
            <View style={styles.signingInformation}>
              <Text style={styles.signText}>
                {strings('signature_request.signing')}
              </Text>
              {showWarning ? (
                <TouchableOpacity
                  style={styles.warningWrapper}
                  onPress={this.goToWarning}
                >
                  <WarningMessage
                    type={'error'}
                    warningMessage={this.renderWarning()}
                  />
                </TouchableOpacity>
              ) : null}
            </View>
            {this.renderActionViewChildren()}
          </View>
        </ActionView>
      </View>
    );
  }

  renderQRDetails() {
    const { QRState, fromAddress } = this.props;
    const styles = this.getStyles();

    return (
      <View style={[styles.root]}>
        <QRSigningDetails
          QRState={QRState}
          showCancelButton
          showHint={false}
          bypassAndroidCameraAccessCheck={false}
          fromAddress={fromAddress}
        />
      </View>
    );
  }

  render() {
    const { isSigningQRObject } = this.props;
    return isSigningQRObject
      ? this.renderQRDetails()
      : this.renderSignatureRequest();
  }
}

const mapStateToProps = (state) => ({
  networkType: state.engine.backgroundState.NetworkController.provider.type,
});

SignatureRequest.contextType = ThemeContext;

export default connect(mapStateToProps)(
  withQRHardwareAwareness(SignatureRequest),
);
