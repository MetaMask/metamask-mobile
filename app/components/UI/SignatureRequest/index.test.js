import React from 'react';
import { shallow } from 'enzyme';
import SignatureRequest from './';
import configureMockStore from 'redux-mock-store';
import { ROPSTEN } from '../../../constants/network';

const mockStore = configureMockStore();

describe('SignatureRequest', () => {
	it('should render correctly', () => {
		const initialState = {
			engine: {
				backgroundState: {
					AccountTrackerController: {
						accounts: { '0x2': { balance: '0' } }
					},
					PreferencesController: {
						selectedAddress: '0x2',
						identities: { '0x2': { address: '0x2', name: 'Account 1' } }
					},
					NetworkController: {
						provider: {
							type: ROPSTEN
						}
					}
				}
			}
		};

		const wrapper = shallow(<SignatureRequest currentPageInformation={{ title: 'title', url: 'url' }} />, {
			context: { store: mockStore(initialState) }
		});
		expect(wrapper.dive()).toMatchSnapshot();
	});
});
