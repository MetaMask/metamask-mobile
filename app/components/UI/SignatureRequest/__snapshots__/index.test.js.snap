// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignatureRequest should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "borderTopLeftRadius": 20,
        "borderTopRightRadius": 20,
        "minHeight": "72%",
        "paddingBottom": 20,
        "paddingTop": 24,
      },
      undefined,
    ]
  }
>
  <Connect(TransactionHeader)
    currentPageInformation={
      Object {
        "title": "title",
        "url": "url",
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "marginVertical": 24,
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "EuclidCircularB-Bold",
          "fontSize": 20,
          "fontWeight": "600",
          "textAlign": "center",
        }
      }
    >
      Sign this message?
    </Text>
  </View>
  <ActionView
    cancelTestID="request-signature-cancel-button"
    cancelText="Cancel"
    confirmButtonMode="sign"
    confirmTestID="request-signature-confirm-button"
    confirmText="Sign"
    confirmed={false}
    onCancelPress={[Function]}
    onConfirmPress={[Function]}
    showCancelButton={true}
    showConfirmButton={true}
  >
    <View
      style={
        Object {
          "paddingHorizontal": 24,
        }
      }
    >
      <View
        style={
          Object {
            "marginBottom": 20,
            "width": "100%",
          }
        }
      >
        <Connect(AccountInfoCard)
          operation="signing"
        />
      </View>
      <TouchableOpacity
        onPress={null}
        style={
          Object {
            "alignSelf": "center",
            "borderColor": "#bbc0c5",
            "borderRadius": 10,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "padding": 16,
            "width": "100%",
          }
        }
      >
        <WebsiteIcon
          style={
            Object {
              "borderRadius": 20,
              "height": 40,
              "marginRight": 8,
              "width": 40,
            }
          }
          title="url"
          url="url"
        />
        <View
          style={
            Object {
              "justifyContent": "space-between",
              "width": "75%",
            }
          }
        >
          <Text
            style={
              Object {
                "fontFamily": "EuclidCircularB-Bold",
                "fontSize": 16,
                "fontWeight": "600",
                "marginBottom": 4,
              }
            }
          >
            Message
            :
          </Text>
        </View>
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flexGrow": 1,
            }
          }
        />
      </TouchableOpacity>
    </View>
  </ActionView>
</View>
`;
